Gems/Atom/Tools/MaterialCanvas/Code/Source/Document/MaterialCanvasDocument.cpp

```diff
@@ -352,4 +372,337 @@ namespace MaterialCanvas
+                for (const auto& slotConfig : nodeConfig.m_outputSlots)
+                {
+                    auto slot = dynamicNode->GetSlot(slotConfig.m_name);
+                    if (slot && !slot->GetConnections().empty())
+                    {
+                        hasOutputConnections = true;
+                        break;
+                    }
+                }
+
+                collectCommonSettings(nodeConfig.m_settings);
+
+                for (const auto& slotConfig : nodeConfig.m_propertySlots)
+                {
+                    collectCommonSettings(slotConfig.m_settings);
+
+                    // Write out any property slot data if there were any connections
+                    if (hasOutputConnections)
+                    {
+                        auto slot = dynamicNode->GetSlot(slotConfig.m_name);
+                        AZStd::vector<AZStd::string> instructionsForSlot;
+                        collectSettingsAsVec(slotConfig.m_settings, "instructions", instructionsForSlot);
+                        updateAndAddSlotInstructions(node, slot, instructionsForSlot);
+                    }
+                }
+
+                for (const auto& slotConfig : nodeConfig.m_inputSlots)
+                {
+                    collectCommonSettings(slotConfig.m_settings);
+
+                    // For now, this will always generate all input slot instructions but can potentially be optimized to eliminate
+                    // instructions that result in temporary variables that are copies would pass through two other variables.
+                    auto slot = dynamicNode->GetSlot(slotConfig.m_name);
+                    AZStd::vector<AZStd::string> instructionsForSlot;
+                    collectSettingsAsVec(slotConfig.m_settings, "instructions", instructionsForSlot);
+
+                    // Input slots will replace the value with the name of the variable from the connected help with slot if one is set.
+                    if (slot && !slot->GetConnections().empty())
+                    {
+                        for (const auto& connection : slot->GetConnections())
+                        {
+                            auto sourceSlot = connection->GetSourceSlot();
+                            if (sourceSlot && sourceSlot->GetSlotDirection() == GraphModel::SlotDirection::Output)
+                            {
+                                replaceStringsInVec(
+                                    "SLOTVALUE",
+                                    AZStd::string::format("node%u_%s", sourceSlot->GetParentNode()->GetId(), sourceSlot->GetName().c_str()),
+                                    instructionsForSlot);
+                                break;
+                            }
+                        }
+                    }
+
+                    updateAndAddSlotInstructions(node, slot, instructionsForSlot);
+                }
+
+                for (const auto& slotConfig : nodeConfig.m_outputSlots)
```

Comment:
Just so I understand, the way I imagine something like this is a loop over the nodes in topological order (as you've done), and potentially processing only input edges. What operations/instructions are associated with "output slots" in particular?

Reply:
The way things are currently set-up, each output slot is defining code that gets generated specifically for it. This is primarily useful in a situation with multiple output slots on the same node generating different code. The settings could be specified on the node instead of the slot to organize things a little differently or to generate a block of instructions or some other data regardless of connections.